### ISSUE TITLE
Window functions over calendar periods

### ISSUE THREAD

**Issue by DavisVaughan** [tidyverts/tsibble#38]
Hi again! Something I've always wanted to get working with `tibbletime`, but couldn't quite figure out (because it wasn't _quite_ flexible enough) was __rolling calculations over "calendar periods" rather than a fixed window__ of say, 5 or 10 units. Playing around and using some of my previous knowledge, I was able to get this kind of thing working with `tsibble`, and I think it is pretty powerful! 

I've created 3 new core functions `islide()`, `itile()` and `istretch()` (and eventually all of their variants) that use the cool `index_by()` idea of "preparing" the `tsibble` for a time based transformation. 

The code all sits in the branch on my page listed below.

I'm interested to hear what you think!

``` r
#devtools::install_github("DavisVaughan/tsibble", ref = "feature/index-window-functions")
library(tsibble)
library(lubridate)
library(dplyr)

pedestrian_bm <- pedestrian %>% filter(Sensor == "Birrarung Marr")

pedestrian_bm_monthly <- pedestrian_bm %>%
  index_by(month = floor_date(Date_Time, "month"))

################################################################################
#### islide()

# islide() performs rolling calculations over calendar periods
# .n determines the number of calendar "chunks" that you roll over (here 1 or 2 months at a time)

# This calculates the mean on months (1) (2) (3) ...
pedestrian_bm_monthly %>%
  islide_dbl(~mean(.x$Count), .n = 1)
#>  [1] 673.1640 501.3705 714.8508 412.8222 423.5069 376.0208 480.5444
#>  [8] 311.7957 402.2125 768.7328 687.1389 471.1556 768.2446 502.6739
#> [15] 649.4046 534.0774 445.9092 336.7472 471.2339 355.0000 391.6250
#> [22] 482.5440 414.5625 502.3575

# ^ Essentially the same as a summarise
pedestrian_bm_monthly %>%
  summarise(x = mean(Count))
#> # A tsibble: 24 x 2 [1HOUR]
#>    month                   x
#>    <dttm>              <dbl>
#>  1 2015-01-01 00:00:00  673.
#>  2 2015-02-01 00:00:00  501.
#>  3 2015-03-01 00:00:00  715.
#>  4 2015-04-01 00:00:00  413.
#>  5 2015-05-01 00:00:00  424.
#>  6 2015-06-01 00:00:00  376.
#>  7 2015-07-01 00:00:00  481.
#>  8 2015-08-01 00:00:00  312.
#>  9 2015-09-01 00:00:00  402.
#> 10 2015-10-01 00:00:00  769.
#> # ... with 14 more rows

# More interesting is being able to roll over arbitrary chunks
# This calculates the mean on months (1,2) (2,3) (3,4) ...
pedestrian_bm_monthly %>%
  islide_dbl(~mean(.x$Count), .n = 2)
#>  [1] 591.6349 566.3122 383.9352 396.1700 458.6334 556.4433 639.8854
#>  [8] 628.1601 389.4461 413.1169 435.4831 497.0366

# islide() works for vectors, but you explicitely pass the .index

val <- pedestrian_bm_monthly$Count
idx <- pedestrian_bm_monthly$month

islide_dbl(val, ~mean(.x), .index = idx, .n = 2)
#>  [1] 591.6349 566.3122 383.9352 396.1700 458.6334 556.4433 639.8854
#>  [8] 628.1601 389.4461 413.1169 435.4831 497.0366

################################################################################
# itile()

# itile() performs a tiling window calculation over non-overlapping calendar periods

# Mean of count over month (1) (2) (3) (4) ...
# In this way, kind of like islide with .n = 1
pedestrian_bm_monthly %>%
  itile_dbl(~mean(.x$Count))
#>  [1] 673.1640 501.3705 714.8508 412.8222 423.5069 376.0208 480.5444
#>  [8] 311.7957 402.2125 768.7328 687.1389 471.1556 768.2446 502.6739
#> [15] 649.4046 534.0774 445.9092 336.7472 471.2339 355.0000 391.6250
#> [22] 482.5440 414.5625 502.3575

################################################################################
# istretch()

# istretch() performs a stretching window calculation over expanding calendar periods

# Calculate the mean over months (1) (1,2) (1,2,3) ...
pedestrian_bm_monthly %>%
  istretch_dbl(~mean(.x$Count))
#>  [1] 673.1640 591.6349 634.0759 578.7625 571.3694 533.8024 524.9735
#>  [8] 494.6592 483.4761 489.6193 502.7165 499.8054 525.0192 523.2143
#> [15] 533.2438 533.2585 527.5043 514.9280 512.1412 502.7196 496.6271
#> [22] 495.9423 495.6602 496.0017
```

Created on 2018-06-24 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

**Comment by DavisVaughan** [comment]
I'm realizing my version of islide() is incorrect. Please leave this open for a few days and I'll correct it, I still think there could be value there. Sorry about that.

**Comment by DavisVaughan** [comment]
I've got `islide()` working how I intended now. I appreciate the clever weighted mean approach to your alternative to `istretch()`, but I'm not sure that generalizes. See the below examples of a linear model rolled over irregular groups of 2 days. Can you do this with the current `tsibble` implementation? I tried a few things but couldn't match it. The idea is almost a generalization of `slide()`, where the `.index` parameter of `islide()` is just `seq_along(.x)` to match what `slide()` does. The generalization allows for rolling over irregular windows, rather than just a fixed window size, which I think is fairly powerful. I'm not sure a good use case has come up for `itile()` though.

Additionally, I think Max Kuhn was interested in having this kind of thing in his `rsample::rolling_origin()` function, to break up data frames over potentially irregular calendar periods rather than just a fixed size. It could be a neat way for `tsibble` to slip in there.

```r
library(tibble)
library(tsibble)
#> 
#> Attaching package: 'tsibble'
#> The following object is masked from 'package:stats':
#> 
#>     filter
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following object is masked from 'package:tsibble':
#> 
#>     id
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union

ex <- tribble(
  ~date, ~var,
"2010-08-04 22:28:07",   867,
"2010-08-04 23:31:12",    891,
"2010-08-04 23:58:05",   901,
"2010-08-05 08:35:50",    991,
"2010-08-05 13:28:02",  1085,
"2010-08-05 14:14:47",   1114,
"2010-08-05 14:21:46",  1117,
"2010-08-05 15:46:24",   1151,
"2010-08-05 16:25:29",  1174,
"2010-08-05 23:19:29",   1268,
"2010-08-06 12:15:42",  1384,
"2010-08-06 15:22:06",   1403,
"2010-08-07 10:25:49",  1550,
"2010-08-07 18:58:16",   1596,
"2010-08-07 21:15:44",  1608
)

ex_ts <- ex %>%
  mutate(date = as.POSIXct(date)) %>%
  as_tsibble(index = date)

ex_ts_day <- ex_ts %>%
  index_by(day = lubridate::floor_date(date, "1 day"))

# This is the idea, obviously the new purrr compliant ones would be more like p-islide()
islide(ex_ts_day, ~lm(var ~ date, data = .), .initial = 2, .skip = 0)
#> [[1]]
#> 
#> Call:
#> lm(formula = var ~ date, data = .)
#> 
#> Coefficients:
#> (Intercept)         date  
#>  -5.682e+06    4.436e-03  
#> 
#> 
#> [[2]]
#> 
#> Call:
#> lm(formula = var ~ date, data = .)
#> 
#> Coefficients:
#> (Intercept)         date  
#>  -4.481e+06    3.499e-03  
#> 
#> 
#> [[3]]
#> 
#> Call:
#> lm(formula = var ~ date, data = .)
#> 
#> Coefficients:
#> (Intercept)         date  
#>  -2.468e+06    1.928e-03

# This is how it separates
# Day (4,5) (5,6) (6,7)
islider(ex_ts_day, .index = ex_ts_day$day, .initial = 2, .skip = 0)
#> [[1]]
#> # A tsibble: 10 x 3 [1SECOND]
#> # Groups:    @ day [2]
#>    date                  var day                
#>    <dttm>              <dbl> <dttm>             
#>  1 2010-08-04 22:28:07   867 2010-08-04 00:00:00
#>  2 2010-08-04 23:31:12   891 2010-08-04 00:00:00
#>  3 2010-08-04 23:58:05   901 2010-08-04 00:00:00
#>  4 2010-08-05 08:35:50   991 2010-08-05 00:00:00
#>  5 2010-08-05 13:28:02  1085 2010-08-05 00:00:00
#>  6 2010-08-05 14:14:47  1114 2010-08-05 00:00:00
#>  7 2010-08-05 14:21:46  1117 2010-08-05 00:00:00
#>  8 2010-08-05 15:46:24  1151 2010-08-05 00:00:00
#>  9 2010-08-05 16:25:29  1174 2010-08-05 00:00:00
#> 10 2010-08-05 23:19:29  1268 2010-08-05 00:00:00
#> 
#> [[2]]
#> # A tsibble: 9 x 3 [1SECOND]
#> # Groups:    @ day [2]
#>   date                  var day                
#>   <dttm>              <dbl> <dttm>             
#> 1 2010-08-05 08:35:50   991 2010-08-05 00:00:00
#> 2 2010-08-05 13:28:02  1085 2010-08-05 00:00:00
#> 3 2010-08-05 14:14:47  1114 2010-08-05 00:00:00
#> 4 2010-08-05 14:21:46  1117 2010-08-05 00:00:00
#> 5 2010-08-05 15:46:24  1151 2010-08-05 00:00:00
#> 6 2010-08-05 16:25:29  1174 2010-08-05 00:00:00
#> 7 2010-08-05 23:19:29  1268 2010-08-05 00:00:00
#> 8 2010-08-06 12:15:42  1384 2010-08-06 00:00:00
#> 9 2010-08-06 15:22:06  1403 2010-08-06 00:00:00
#> 
#> [[3]]
#> # A tsibble: 5 x 3 [1SECOND]
#> # Groups:    @ day [2]
#>   date                  var day                
#>   <dttm>              <dbl> <dttm>             
#> 1 2010-08-06 12:15:42  1384 2010-08-06 00:00:00
#> 2 2010-08-06 15:22:06  1403 2010-08-06 00:00:00
#> 3 2010-08-07 10:25:49  1550 2010-08-07 00:00:00
#> 4 2010-08-07 18:58:16  1596 2010-08-07 00:00:00
#> 5 2010-08-07 21:15:44  1608 2010-08-07 00:00:00

# lm on (4) (4,5) (4,5,6) (4,5,6,7) 
istretch(ex_ts_day, ~lm(var ~ date, data = .))
#> [[1]]
#> 
#> Call:
#> lm(formula = var ~ date, data = .)
#> 
#> Coefficients:
#> (Intercept)         date  
#>  -8.077e+06    6.306e-03  
#> 
#> 
#> [[2]]
#> 
#> Call:
#> lm(formula = var ~ date, data = .)
#> 
#> Coefficients:
#> (Intercept)         date  
#>  -5.682e+06    4.436e-03  
#> 
#> 
#> [[3]]
#> 
#> Call:
#> lm(formula = var ~ date, data = .)
#> 
#> Coefficients:
#> (Intercept)         date  
#>  -4.818e+06    3.762e-03  
#> 
#> 
#> [[4]]
#> 
#> Call:
#> lm(formula = var ~ date, data = .)
#> 
#> Coefficients:
#> (Intercept)         date  
#>  -3.703e+06    2.892e-03

```

**Comment by earowang** [comment]
Thanks, I've realised that `lslide()`/`ltile()`/`lstretch()` handling list inputs return unexpected results, now they've been fixed. The approach is to use `split_by()` coupled with `lslide()`, which is more transparent I think.

``` r
library(tsibble)
#> 
#> Attaching package: 'tsibble'
#> The following object is masked from 'package:stats':
#> 
#>     filter
ex <- tibble::tribble(
  ~date, ~var,
  "2010-08-04 22:28:07",   867,
  "2010-08-04 23:31:12",    891,
  "2010-08-04 23:58:05",   901,
  "2010-08-05 08:35:50",    991,
  "2010-08-05 13:28:02",  1085,
  "2010-08-05 14:14:47",   1114,
  "2010-08-05 14:21:46",  1117,
  "2010-08-05 15:46:24",   1151,
  "2010-08-05 16:25:29",  1174,
  "2010-08-05 23:19:29",   1268,
  "2010-08-06 12:15:42",  1384,
  "2010-08-06 15:22:06",   1403,
  "2010-08-07 10:25:49",  1550,
  "2010-08-07 18:58:16",   1596,
  "2010-08-07 21:15:44",  1608
)

ex_ts <- ex %>%
  mutate(date = as.POSIXct(date)) %>%
  as_tsibble(index = date)

ex_ts_day <- ex_ts %>%
  mutate(day = lubridate::floor_date(date, "1 day")) %>% 
  split_by(day)

ex_ts_day %>% 
  lslide(~ lm(var ~ date, data = .), .size = 2)
#> [[1]]
#> [1] NA
#> 
#> [[2]]
#> 
#> Call:
#> lm(formula = var ~ date, data = .)
#> 
#> Coefficients:
#> (Intercept)         date  
#>  -5.681e+06    4.436e-03  
#> 
#> 
#> [[3]]
#> 
#> Call:
#> lm(formula = var ~ date, data = .)
#> 
#> Coefficients:
#> (Intercept)         date  
#>  -4.481e+06    3.499e-03  
#> 
#> 
#> [[4]]
#> 
#> Call:
#> lm(formula = var ~ date, data = .)
#> 
#> Coefficients:
#> (Intercept)         date  
#>  -2.468e+06    1.928e-03
ex_ts_day %>% 
  lstretch(~ lm(var ~ date, data = .), .size = 1)
#> [[1]]
#> 
#> Call:
#> lm(formula = var ~ date, data = .)
#> 
#> Coefficients:
#> (Intercept)         date  
#>  -8.077e+06    6.306e-03  
#> 
#> 
#> [[2]]
#> 
#> Call:
#> lm(formula = var ~ date, data = .)
#> 
#> Coefficients:
#> (Intercept)         date  
#>  -5.681e+06    4.436e-03  
#> 
#> 
#> [[3]]
#> 
#> Call:
#> lm(formula = var ~ date, data = .)
#> 
#> Coefficients:
#> (Intercept)         date  
#>  -4.818e+06    3.762e-03  
#> 
#> 
#> [[4]]
#> 
#> Call:
#> lm(formula = var ~ date, data = .)
#> 
#> Coefficients:
#> (Intercept)         date  
#>  -3.703e+06    2.892e-03
```

Created on 2018-07-01 by the [reprex
package](http://reprex.tidyverse.org) (v0.2.0).

**Comment by DavisVaughan** [comment]
I think this is definitely a move in the right direction. One last alarm is going off in my head. I'm gonna apologize in advance for being super nit-picky, I promise I only have the best intentions in mind, as I really like where this is heading!

When I think of what I would use after the `split_by` in a `purrr` context, it would be `map` not `lmap`. `map` can iterate over lists, but currently `slide` and friends do not. I think your idea is that `lslide` replaces the idea of `slide` iterating over lists, but I'm currently thinking that might be "theoretically" inconsistent with the `purrr` definition of `lmap`. 

From `purrr` docs) _Thus, instead of mapping the elements of a list (as in .x[[i]]), they apply a function .f to each subset of size 1 of that list (as in .x[i]). We call those those elements list-elements)._

To show the inconsistency:

```r
library(dplyr)

ex <- tibble::tribble(
  ~date, ~var,
  "2010-08-04 22:28:07",   867,
  "2010-08-04 23:31:12",    891,
  "2010-08-04 23:58:05",   901,
  "2010-08-05 08:35:50",    991,
  "2010-08-05 13:28:02",  1085,
  "2010-08-05 14:14:47",   1114,
  "2010-08-05 14:21:46",  1117,
  "2010-08-05 15:46:24",   1151,
  "2010-08-05 16:25:29",  1174,
  "2010-08-05 23:19:29",   1268,
  "2010-08-06 12:15:42",  1384,
  "2010-08-06 15:22:06",   1403,
  "2010-08-07 10:25:49",  1550,
  "2010-08-07 18:58:16",   1596,
  "2010-08-07 21:15:44",  1608
)

ex_ts_day <- ex %>%
  mutate(date = as.POSIXct(date)) %>%
  as_tsibble(index = date) %>%
  mutate(day = lubridate::floor_date(date, "1 day"))

# List elements .x[i]
ex_ts_day %>%
  split_by(day) %>%
  purrr::lmap(~list(length(.)))
#> [[1]]
#> [1] 1
#> 
#> [[2]]
#> [1] 1
#> 
#> [[3]]
#> [1] 1
#> 
#> [[4]]
#> [1] 1

# The actual thing in the list .x[[i]] 
ex_ts_day %>%
  split_by(day) %>%
  purrr::map(~length(.))
#> [[1]]
#> [1] 3
#> 
#> [[2]]
#> [1] 3
#> 
#> [[3]]
#> [1] 3
#> 
#> [[4]]
#> [1] 3

# lslide is acting like map(), not lmap()
ex_ts_day %>%
  split_by(day) %>%
  lslide(~length(.), .size = 2)
#> [[1]]
#> [1] NA
#> 
#> [[2]]
#> [1] 3
#> 
#> [[3]]
#> [1] 3
#> 
#> [[4]]
#> [1] 3
```

See how `lmap` is acting only on `.x[i]` so its length is 1, but for `lslide` it acts on `.x[[i]]` so the length is the number of cols? It's acting more like `map()` in that case.

Personally, I think `slide()` could work with lists (and data frames). This has a few benefits:

1) Get access to `slide_dbl` and friends
2) More people know `map` over `lmap` and would be more comfortable with `slide` iterating over lists (honestly my main use case of `map` is over lists so that feels more natural to me too)
3) Could adjust `lslide` to work consistently with `lmap`
4) Get access to `slider()` working correctly over lists and data frames, which could be really useful.

I currently have the basic idea working, stealing a few bits from `list_constructor()` and adjusting `slider()`. I also changed the use of `bind_rows` in `list_constructor` to a version of that that preserves `tsibble` attributes, since that's generally a problem with that non-generic function (I had the same issues with `tibbletime`). It might be able to be cleaned up a bit (not sure how much I like relying on `vec_depth()` of 4) but overall seems to work well.

To me, the below workflow is super intuitive. I especially like the `tidyr::nest()` bit, since you get to keep the index that you are sliding along!

``` r

# devtools::install_github("DavisVaughan/tsibble", ref = "feature/window-function-with-lists")

library(tsibble)
library(dplyr)

ex <- tibble::tribble(
  ~date, ~var,
  "2010-08-04 22:28:07",   867, "2010-08-04 23:31:12",    891, "2010-08-04 23:58:05",   901, "2010-08-05 08:35:50",    991,
  "2010-08-05 13:28:02",  1085, "2010-08-05 14:14:47",   1114, "2010-08-05 14:21:46",  1117, "2010-08-05 15:46:24",   1151,
  "2010-08-05 16:25:29",  1174, "2010-08-05 23:19:29",   1268, "2010-08-06 12:15:42",  1384, "2010-08-06 15:22:06",   1403,
  "2010-08-07 10:25:49",  1550, "2010-08-07 18:58:16",   1596, "2010-08-07 21:15:44",  1608
)

ex_ts_day <- ex %>%
  mutate(date = as.POSIXct(date)) %>%
  as_tsibble(index = date) %>%
  mutate(day = lubridate::floor_date(date, "1 day"))

ex_ts_day %>%
  split_by(day) %>%
  slide(~nrow(.), .size = 2)
#> [[1]]
#> [1] NA
#> 
#> [[2]]
#> [1] 10
#> 
#> [[3]]
#> [1] 9
#> 
#> [[4]]
#> [1] 5

ex_ts_day %>%
  tidyr::nest(-day) %>%
  mutate(n = slide_dbl(data, ~nrow(.), .size = 2))
#> # A tibble: 4 x 3
#>   day                 data                  n
#>   <dttm>              <list>            <dbl>
#> 1 2010-08-04 00:00:00 <tsibble [3 × 2]>    NA
#> 2 2010-08-05 00:00:00 <tsibble [7 × 2]>    10
#> 3 2010-08-06 00:00:00 <tsibble [2 × 2]>     9
#> 4 2010-08-07 00:00:00 <tsibble [3 × 2]>     5

ex_ts_day %>%
  split_by(day) %>%
  slider(.size = 2)
#> [[1]]
#> # A tsibble: 10 x 3 [1SECOND]
#>    date                  var day                
#>    <dttm>              <dbl> <dttm>             
#>  1 2010-08-04 22:28:07   867 2010-08-04 00:00:00
#>  2 2010-08-04 23:31:12   891 2010-08-04 00:00:00
#>  3 2010-08-04 23:58:05   901 2010-08-04 00:00:00
#>  4 2010-08-05 08:35:50   991 2010-08-05 00:00:00
#>  5 2010-08-05 13:28:02  1085 2010-08-05 00:00:00
#>  6 2010-08-05 14:14:47  1114 2010-08-05 00:00:00
#>  7 2010-08-05 14:21:46  1117 2010-08-05 00:00:00
#>  8 2010-08-05 15:46:24  1151 2010-08-05 00:00:00
#>  9 2010-08-05 16:25:29  1174 2010-08-05 00:00:00
#> 10 2010-08-05 23:19:29  1268 2010-08-05 00:00:00
#> 
#> [[2]]
#> # A tsibble: 9 x 3 [1SECOND]
#>   date                  var day                
#>   <dttm>              <dbl> <dttm>             
#> 1 2010-08-05 08:35:50   991 2010-08-05 00:00:00
#> 2 2010-08-05 13:28:02  1085 2010-08-05 00:00:00
#> 3 2010-08-05 14:14:47  1114 2010-08-05 00:00:00
#> 4 2010-08-05 14:21:46  1117 2010-08-05 00:00:00
#> 5 2010-08-05 15:46:24  1151 2010-08-05 00:00:00
#> 6 2010-08-05 16:25:29  1174 2010-08-05 00:00:00
#> 7 2010-08-05 23:19:29  1268 2010-08-05 00:00:00
#> 8 2010-08-06 12:15:42  1384 2010-08-06 00:00:00
#> 9 2010-08-06 15:22:06  1403 2010-08-06 00:00:00
#> 
#> [[3]]
#> # A tsibble: 5 x 3 [1SECOND]
#>   date                  var day                
#>   <dttm>              <dbl> <dttm>             
#> 1 2010-08-06 12:15:42  1384 2010-08-06 00:00:00
#> 2 2010-08-06 15:22:06  1403 2010-08-06 00:00:00
#> 3 2010-08-07 10:25:49  1550 2010-08-07 00:00:00
#> 4 2010-08-07 18:58:16  1596 2010-08-07 00:00:00
#> 5 2010-08-07 21:15:44  1608 2010-08-07 00:00:00
```

**Comment by DavisVaughan** [comment]
After much thinking of how these functions should ideally behave, I've rewritten slide/stretch/tile to all "do less" to actually end up accomplishing much more in a more consistent manner. I'll post the code tomorrow to get your input. I wouldn't move forward with any of the above suggestions at the moment if you had plans to.

**Comment by DavisVaughan** [comment]
I've rewritten `slide()`, `stretch()`, `tile()` and all of their friends. There are also new `slider2()` and `pslider()` functions along with equivalents for `tiler()` and `stretcher()`. 

Everything takes lists and data frames now. There is much less attempt to auto-simplify anything based on the vector depth. 

I'll explain much more in detail tomorrow on my thoughts of why things now work the way they do, but in the mean time the code (without updated docs or tests!) is on the branch linked below

```r
devtools::install_github("DavisVaughan/tsibble", ref = "feature/window-function-rewrite")
```

There are a ton of examples here:

https://gist.github.com/DavisVaughan/b3f80fe0c71a5f3cd056d7a6da12207e

One separate issue to account for (somehow) is that `bind_rows()` breaks when binding 2 tsibble objects. It isn't generic so you can't easily fix it with a new method. This was an (unsolved) problem with tibbletime too. It will come into play when rolling over calendar periods.

**Comment by earowang** [comment]
It seems to me that adding a new argument to the current `slider()` could do what `slider2()` and `pslider()` do. When a list is passed to `slider()`, `slider()` needs to distinguish it should slide over a list, or slide over the elements within a list simultaneously, introducing an argument could suffice.

```r
slider2(.x, .y, .size = 2)
#> [[1]]
#> [[1]][[1]]
#> [1] 1 2
#> 
#> [[1]][[2]]
#> [1] 6 7
#> 
#> 
#> [[2]]
#> [[2]][[1]]
#> [1] 2 3
#> 
#> [[2]][[2]]
#> [1] 7 8
#> 
#> 
#> [[3]]
#> [[3]][[1]]
#> [1] 3 4
#> 
#> [[3]][[2]]
#> [1] 8 9
#> 
#> 
#> [[4]]
#> [[4]][[1]]
#> [1] 4 5
#> 
#> [[4]][[2]]
#> [1]  9 10
```

I would expect that `slider2(.x, .y, .size = 2)` returns a list of length 2 and each element returns a list of length 4.

**Comment by DavisVaughan** [comment]
"needs to distinguish it should slide over a list, or slide over the elements within a list simultaneously" I think this is definitely right.

Can you explain further what you mean by:
"I would expect that slider2(.x, .y, .size = 2) returns a list of length 2 and each element returns a list of length 4."
I'm trying to picture it in my head but can't.

When thinking about `slider2()`, I thought it lined up with `map2()`.

``` r

.x <- 1:3
.y <- 6:8

tsibble::slider2(.x, .y, .size = 1)
#> [[1]]
#> [[1]][[1]]
#> [1] 1
#> 
#> [[1]][[2]]
#> [1] 6
#> 
#> 
#> [[2]]
#> [[2]][[1]]
#> [1] 2
#> 
#> [[2]][[2]]
#> [1] 7
#> 
#> 
#> [[3]]
#> [[3]][[1]]
#> [1] 3
#> 
#> [[3]][[2]]
#> [1] 8

purrr::map2(.x, .y, ~list(.x, .y))
#> [[1]]
#> [[1]][[1]]
#> [1] 1
#> 
#> [[1]][[2]]
#> [1] 6
#> 
#> 
#> [[2]]
#> [[2]][[1]]
#> [1] 2
#> 
#> [[2]][[2]]
#> [1] 7
#> 
#> 
#> [[3]]
#> [[3]][[1]]
#> [1] 3
#> 
#> [[3]][[2]]
#> [1] 8
```

**Comment by DavisVaughan** [comment]
Okay I've spent a bit of time really digging into this and the potential differences between our approaches. You should be able to open the HTML in the attached zip. It's all explained there.

If you're low on time, you may want to start on "Where might we disagree?"
The sections on map() and pmap() are also good ones to see my thoughts explained as well.

[tsibble-window.zip](https://github.com/tidyverts/tsibble/files/2161448/tsibble-window.zip)

**Comment by earowang** [comment]
I've restructured quite a lot on `slide()` this morning. I figure we can see eye to eye on this matter now.

**Comment by DavisVaughan** [comment]
I think with the most recent changes we now agree on what `slider()` should do. I think you can actually simplify `slider()` further to just be 

```r
slider <- function(.x, .size = 1) {
 if (is.data.frame(.x)) 
    .x <- as.list(.x)
 return(slider_base(.x, .size = .size))
}
```
__I think there is still some disagreement on `pslider()`:__

With `pslider()`, the main thing I want to do is pass it a list and have it iterate over the elements of that list in parallel and return that result to me. I currently can't do this. The below returns the same output as `slider()` (with one more layer of list hierarchy), so I think there is overlap in responsibilities.

```r
library(tsibble)
.x <- 1:5
.y <- 6:10
.z <- 11:15
.lst <- list(x = .x, y = .y, z = .z)
.df <- as.data.frame(.lst)

pslider(.lst, .size = 2)
# [[1]]
# [[1]][[1]]
# [[1]][[1]]$x
# [1] 1 2 3 4 5
# 
# [[1]][[1]]$y
# [1]  6  7  8  9 10
# 
# 
# [[1]][[2]]
# [[1]][[2]]$y
# [1]  6  7  8  9 10
# 
# [[1]][[2]]$z
# [1] 11 12 13 14 15
```

What I expect:

```r
pslider(.lst, .size = 2)
# [[1]]
# [[1]][[1]]
# [1] 1 2
# 
# [[1]][[2]]
# [1] 6 7
# 
# [[1]][[3]]
# [1] 11 12
# 
# 
# [[2]]
# [[2]][[1]]
# [1] 2 3
# 
# [[2]][[2]]
# [1] 7 8
# 
# [[2]][[3]]
# [1] 12 13
# 
# 
# [[3]]
# [[3]][[1]]
# [1] 3 4
# 
# [[3]][[2]]
# [1] 8 9
# 
# [[3]][[3]]
# [1] 13 14
# 
# 
# [[4]]
# [[4]][[1]]
# [1] 4 5
# 
# [[4]][[2]]
# [1]  9 10
# 
# [[4]][[3]]
# [1] 14 15
```

With the current implementation, the only way to get this is:

```r
pslider(.x, .y, .z, .size = 2)
```

which feels wrong when compared to what `pmap()` does by only accepting lists as input. I just don't think `pslider()` is living up to its name by not iterating over that list in parallel, and by accepting things other than lists. If you wanted to pass in something other than a list, you use `slider()`. And again, people often want to iterate over 2 things in parallel, which is the case for `slider2()` as a shortcut for `pslider(list(x,y), .size = 2)`

_This also makes sense because when you use `pslide()`, it returns the results in the way I expect them._ I use this for debugging my `pslide()` calls sometimes. Try this and note the difference in how things are returned, then look at what I was expecting from `pslider()` with a list and compare to `pslide()` below:

```r
pslide(list(.x, .y, .z), ~list(.x, .y, ..3), .size = 2)

pslider(list(.x, .y, .z), .size = 2)
```

__Also `pslider()` should not care about depth__

I think the following is perfectly valid based on my implementation of `pslider()`:

```r
.df %>%
  as_tibble() %>%
  select(x) %>%
  slice(1:3) %>%
  mutate(xx = purrr::map(x, ~c(.x, .x, .x))) %>%
  pslider(.size = 2)

# [[1]]
# [[1]][[1]]
# [1] 1 2
# 
# [[1]][[2]]
# [[1]][[2]][[1]]
# [1] 1 1 1
# 
# [[1]][[2]][[2]]
# [1] 2 2 2
# 
# 
# 
# [[2]]
# [[2]][[1]]
# [1] 2 3
# 
# [[2]][[2]]
# [[2]][[2]][[1]]
# [1] 2 2 2
# 
# [[2]][[2]][[2]]
# [1] 3 3 3
```

The current version errors with `"must not be deeper than 3"` because of the nested list, which I think a lot of people might have in some cases.

__And on `slide2()`:__

I still don't think `slide2()` should ever attempt to simplify it's input. `tsibble` currently does so if the first argument is a list.

```r
slide2(.lst, .x[1:3], ~list(.x, .y), .size = 2)

# [[1]]
# [1] NA
# 
# [[2]]
# [[2]][[1]]
# [1]  1  2  3  4  5  6  7  8  9 10
# 
# [[2]][[2]]
# [1] 1 2
# 
# 
# [[3]]
# [[3]][[1]]
# [1]  6  7  8  9 10 11 12 13 14 15
# 
# [[3]][[2]]
# [1] 2 3
```

I would have expected:

```r
slide2(.lst, .x[1:3], ~list(.x, .y), .size = 2)

# [[1]]
# [1] NA
# 
# [[2]]
# [[2]][[1]]
# [[2]][[1]]$x
# [1] 1 2 3 4 5
# 
# [[2]][[1]]$y
# [1]  6  7  8  9 10
# 
# 
# [[2]][[2]]
# [1] 1 2
# 
# 
# [[3]]
# [[3]][[1]]
# [[3]][[1]]$y
# [1]  6  7  8  9 10
# 
# [[3]][[1]]$z
# [1] 11 12 13 14 15
# 
# 
# [[3]][[2]]
# [1] 2 3
```

What if the elements of the list are not all numeric? The simplification causes potential issues:

```r
slide2(list("A", c(1,2), "B"), .x[1:3], ~list(.x, .y), .size = 2)

# [[1]]
# [1] NA
# 
# [[2]]
# [[2]][[1]]
# [1] "A" "1" "2"
# 
# [[2]][[2]]
# [1] 1 2
# 
# 
# [[3]]
# [[3]][[1]]
# [1] "1" "2" "B"
# 
# [[3]][[2]]
# [1] 2 3
```

VS what I expect:

```r
slide2(list("A", c(1,2), "B"), .x[1:3], ~list(.x, .y), .size = 2)

[[1]]
[1] NA

[[2]]
[[2]][[1]]
[[2]][[1]][[1]]
[1] "A"

[[2]][[1]][[2]]
[1] 1 2


[[2]][[2]]
[1] 1 2


[[3]]
[[3]][[1]]
[[3]][[1]][[1]]
[1] 1 2

[[3]][[1]][[2]]
[1] "B"


[[3]][[2]]
[1] 2 3
```

**Comment by DavisVaughan** [comment]
Out of curiosity, was there something in particular you didn't like about my rewrite? It seems to be where we are moving, and I'm pretty confident about the implementation

**Comment by DavisVaughan** [comment]
I think I agree with 99% of the changes in your last commit, and it makes things simpler. 

I can get on board with how `pslider()` works when I look at it more as a helper function.

The 1% I'm unsure about is a new thing I noticed. The backfilling of `NA` values is now moved inside slider. Are you sure about this? It now kind of supports this idea of a partial slide, which might be more headache than anything else. I liked the idea of applying that `.f` function to only complete sliding slices of data, and then adding on the `NA` values later, now it applies it to slices where the first slice has an `NA` value (something not in the user's original source of data), which can cause issues. See below:

``` r
library(tsibble)
#> 
#> Attaching package: 'tsibble'
#> The following object is masked from 'package:stats':
#> 
#>     filter
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following object is masked from 'package:tsibble':
#> 
#>     id
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(purrr)

.x <- 1:5
.y <- 6:10
.z <- 11:15
.lst <- list(x = .x, y = .y, z = .z)
.df <- as.data.frame(.lst)

# Rolling mean, things look normal because mean(c(NA, 1)) is NA
.df %>%
  mutate(x_roll_mean = slide(x, ~mean(.x), .size = 2))
#>   x  y  z x_roll_mean
#> 1 1  6 11          NA
#> 2 2  7 12         1.5
#> 3 3  8 13         2.5
#> 4 4  9 14         3.5
#> 5 5 10 15         4.5

# What if I was trying to be safe and ignore some potential NA value problems
# that could occur down the line in my data set?
# I get 1 as the first value, likely not what I expected!!!
.df %>%
  mutate(x_roll_mean = slide(x, ~mean(.x, na.rm = TRUE), .size = 2))
#>   x  y  z x_roll_mean
#> 1 1  6 11           1
#> 2 2  7 12         1.5
#> 3 3  8 13         2.5
#> 4 4  9 14         3.5
#> 5 5 10 15         4.5


# Also, how would you do this now? You gave that row oriented example
# but i don't see how it lines up to this.
# You get bind_rows(list(NA, <tibble>)) which fails.

nested_monthly <- pedestrian %>%
  index_by(mon = yearmonth(Date_Time)) %>%
  nest(-mon)

nested_monthly %>%
  # Work around for bind_rows for tsibbles not working
  mutate(data = map(data, as_tibble)) %>%
  # .x is list(data[1:2]) and so on. Again, predictable output
  mutate(rolling_mod = slide(data, ~ lm(Count ~ Time + Sensor, data = bind_rows(.x)), .size = 2))
#> Error in mutate_impl(.data, dots): Evaluation error: Argument 1 must have names.
```

Created on 2018-07-05 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

When I do these rolling calculations with `.size = 2`, I typically think of it in complete sets as: (1, 2) (2, 3) not as the partial set of (NA, 1) (1, 2) (2, 3). Does that make sense?

Also, I liked how `slider()` split things into complete rolling slices, and then `slide()` back filled with `NA` separately to work nicely with `mutate()`. `slider()` could be useful in other situations than with `mutate()` and I wouldn't think the user would want the `NA` values there.

Besides that, I think I am on board with everything else. Thanks for being patient with me! It's been fun to go back and forth with you on this.

**Comment by earowang** [comment]
To accomodate different expectations, `slide()` gains a new argument `.partial = FALSE` by default to do partial sliding or not.

Slidling over 2 months:

``` r
library(tsibble)
#> 
#> Attaching package: 'tsibble'
#> The following object is masked from 'package:stats':
#> 
#>     filter
nested_monthly <- pedestrian %>%
  index_by(mon = yearmonth(Date_Time)) %>%
  nest(-mon)

nested_monthly %>%
  # mutate(data = map(data, as_tibble)) %>% # no need
  mutate(rolling_mod = slide(data, ~ lm(Count ~ Time + Sensor, data = dplyr::bind_rows(.x)), .size = 2))
#> # A tibble: 24 x 3
#>         mon data                  rolling_mod
#>       <mth> <list>                <list>     
#>  1 2015 Jan <tsibble [2,232 × 5]> <lgl [1]>  
#>  2 2015 Feb <tsibble [2,304 × 5]> <S3: lm>   
#>  3 2015 Mar <tsibble [2,976 × 5]> <S3: lm>   
#>  4 2015 Apr <tsibble [2,880 × 5]> <S3: lm>   
#>  5 2015 May <tsibble [2,376 × 5]> <S3: lm>   
#>  6 2015 Jun <tsibble [2,880 × 5]> <S3: lm>   
#>  7 2015 Jul <tsibble [2,976 × 5]> <S3: lm>   
#>  8 2015 Aug <tsibble [2,976 × 5]> <S3: lm>   
#>  9 2015 Sep <tsibble [2,880 × 5]> <S3: lm>   
#> 10 2015 Oct <tsibble [2,360 × 5]> <S3: lm>   
#> # ... with 14 more rows
```

Created on 2018-07-06 by the [reprex
package](http://reprex.tidyverse.org) (v0.2.0).

**Comment by DavisVaughan** [comment]
Ah, great idea! Okay, I'm happy with how all of this is working now.

(PS I think you need `.partial = .partial` here since `pslider()` takes `...`)
https://github.com/tidyverts/tsibble/blob/master/R/slide.R#L114

Is this supposed to be an error?

```r
.x <- 1:10
slide(.x, ~.x, .size = -2, .partial = FALSE)
#Error in rep_len(.fill, .size - 1) : invalid 'length.out' value
```

**Comment by DavisVaughan** [comment]
Also I see there is no need for the conversion to `tibble` in that example above, but just want to make sure you are aware of the bug generally. Not sure if there is anything to do besides wait for a better `dplyr` solution (many people have complained about how bind_rows() works in this way)

```r
dplyr::bind_rows(pedestrian, pedestrian)
#Error: The `index` has been dropped somehow. Please reconstruct the `tbl_ts`.
```

**Comment by earowang** [comment]
Many dplyr verbs don't protect attributes for other objects. `bind_rows()` isn't an S3 generic, I can't do much about it. But `rbind(pedestrian, pedestrian)` always works.

**Comment by earowang** [comment]
Thanks for catching those potential issues on the code.
